research/delf/delf/python/training/global_features_utils.py

```diff
@@ -0,0 +1,212 @@
+                                            medium_ground_truth,
+                                            desired_pr_ranks)
+    metrics_hard = dataset.ComputeMetrics(sorted_index_ids, hard_ground_truth,
+                                          desired_pr_ranks)
+
+    debug_and_log(
+      '>> {}: mAP E: {}, M: {}, H: {}'.format(
+        dataset_name, np.around(metrics_easy[0] * 100, decimals=2),
+        np.around(metrics_medium[0] * 100, decimals=2),
+        np.around(metrics_hard[0] * 100, decimals=2)), log=log)
+
+    debug_and_log(
+      '>> {}: mP@k{} E: {}, M: {}, H: {}'.format(
+        dataset_name, desired_pr_ranks,
+        np.around(metrics_easy[1] * 100, decimals=2),
+        np.around(metrics_medium[1] * 100, decimals=2),
+        np.around(metrics_hard[1] * 100, decimals=2)), log=log)
+
+    return metrics_easy, metrics_medium, metrics_hard
+
+
+def htime(time_difference):
+  """Time formatting function.
+
+  Depending on the value of `time_difference` outputs time in an appropriate
+  time format.
+
+  Args:
+    time_difference: Float, time difference between the two events.
+
+  Returns:
+    time: String representing time in an appropriate time format.
+  """
+  time_difference = round(time_difference)
+
+  days = time_difference // 86400
+  hours = time_difference // 3600 % 24
+  minutes = time_difference // 60 % 60
+  seconds = time_difference % 60
+
+  if days > 0:
+    return '{:d}d {:d}h {:d}m {:d}s'.format(days, hours, minutes, seconds)
+  if hours > 0:
+    return '{:d}h {:d}m {:d}s'.format(hours, minutes, seconds)
+  if minutes > 0:
+    return '{:d}m {:d}s'.format(minutes, seconds)
+  return '{:d}s'.format(seconds)
+
+
+def debug_and_log(msg, debug=True, log=True, debug_on_the_same_line=False):
+  """Outputs `msg` to both stdout (if in the debug mode) and the log file.
+
+  Args:
+    msg: String, message to be logged.
+    debug: Bool, if True, will print `msg` to stdout.
+    log: Bool, if True, will redirect `msg` to the logfile.
+    debug_on_the_same_line: Bool, if True, will print `msg` to stdout without
+      a new line. When using this mode, logging to a logfile is disabled.
+  """
+  if debug_on_the_same_line:
+    print(msg, end='')
+    return
+  if debug:
+    print(msg)
+  if log:
+    logging.info(msg)
+
+
+def launch_tensorboard(log_dir):
+  """Runs tensorboard with the given `log_dir` and waits for user input to kill
+  the app.
+  
+  Args:
+    log_dir: String, directory to start tensorboard in.
+  """
+  tb = program.TensorBoard()
+  tb.configure(argv=[None, '--logdir', log_dir])
+  url = tb.launch()
+  debug_and_log("Launching Tensorboard: {}".format(url))
+
+
+def get_standard_keras_models():
+  """Gets the standard keras model names.
+
+  Returns:
+    model_names: List, names of the standard keras models.
+  """
+  model_names = sorted(name for name in tf.keras.applications.__dict__ if not
+  name.startswith("__") and callable(tf.keras.applications.__dict__[name]))
```

Comment:
pls format this according to the style guide: https://google.github.io/styleguide/pyguide.html#211-conditional-expressions